### PR TITLE
Enable returning BlobRefs in the API, enable tests, and minor fixes

### PR DESCRIPTION
**Pull request by dcoutts**
This should now complete the functionality and testing for the blobs feature.

Previous PRs changed the way blobs were managed in the write buffer. This PR does the final wiring between lookups in the write buffer and returning BlobRefs via the public API.

Then it also enables generating blobs and blobrefs in various tests. In particular this includes the state machine tests. Enabling these uncovered a few mismatches between the model and implementation. See issue #392.

It also covers some of the overlap between cursors and blobs: returning blob references from cursors is now enabled and the tests are passing.

There are sill some TODOs that would be nice to do. In particular the way the write buffer is handled during lookup is rather ugly. The use of the write buffer could be properly integrated into `lookupIO`. And related, there's more allocations than necessary in various conversions in the lookup process, and part of that is to do with blob references.

### CODE REVIEW

**Comment by dcoutts** [comment]
This will need a rebase once #389 is merged.

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 1]
```diff
@@ -1082,9 +1083,10 @@ arbitraryActionWithVars _ findVars _st = QC.frequency $ concat [
     genDeletes :: Gen (V.Vector k)
     genDeletes = QC.arbitrary
 
-    -- TODO: generate @Just blob@ once blob references are implemented
     genBlob :: Gen (Maybe blob)
-    genBlob = Nothing <$ QC.arbitrary @blob
+    genBlob = QC.oneof [ pure Nothing
+                       , Just <$> QC.arbitrary @blob
+                       ]
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:1089]: Isn't this just `QC.arbitrary @(Maybe blob)`?

**Review by mheinzel — COMMENTED** [review]
Looks good, just one potential issue with cleaning up the `WriteBufferBlobs` used for cursors.

File: test/Test/Database/LSMTree/Normal/StateMachine.hs (hunk 1 above)
Comment (dcoutts) [reply test/Test/Database/LSMTree/Normal/StateMachine.hs:1089]: Um yes. :grin:

File: src/Database/LSMTree/Internal/WriteBufferBlobs.hs [hunk 2]
```diff
@@ -226,25 +226,27 @@ readBlob fs WriteBufferBlobs {blobFileState} BlobSpan {blobSpanOffset, blobSpanS
...
 -- | Helper function to make a 'BlobRef' that points into a 'WriteBufferBlobs'.
+-- TODO: this two stage function design isn't terribly nice, and allocates
+-- more than necessary (a closure every call).
 mkBlobRef :: PrimMonad m
           => WriteBufferBlobs m h
-          -> BlobSpan
-          -> m (BlobRef m (FS.Handle h))
-mkBlobRef WriteBufferBlobs {blobFileState, blobFileRefCounter} blobRefSpan = do
+          -> m (BlobSpan -> BlobRef m (FS.Handle h))
+mkBlobRef WriteBufferBlobs {blobFileState, blobFileRefCounter} = do
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/WriteBufferBlobs.hs:234]: In one place where it's used, it's much more convenient for the conversion to be a pure function.

But I think the whole set of conversions could be done better. So I thought it might be better to resolve it as part of doing that.

File: src/Database/LSMTree/Internal.hs [hunk 3]
```diff
@@ -868,12 +877,17 @@ newCursor th = withOpenTable th $ \thEnv -> do
     -- references to each run, so it is safe.
     allocTableContent reg contentVar = do
         RW.withReadAccess contentVar $ \content -> do
+          let wb      = tableWriteBuffer content
+              wbblobs = tableWriteBufferBlobs content
+          allocateTemp reg
+            (WBB.addReference wbblobs)
+            (\_ -> WBB.removeReference wbblobs)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:889]: Ok, so if I've understood our temp registry properly, then the code above will add the reference, and then it will only remove the reference if an exception is thrown subsequently but within the scope of `withTempRegistry`. So in other words this will deal with the allocation and failure case, but not the happy case.

So presumably for that we should include the wbblobe into the cursor itself, and release it in `closeCursor`.

[the remark this answers is not in the document]
Comment (mheinzel) [reply src/Database/LSMTree/Internal.hs:894]: Exactly!

[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:894]: I applied a fix, and it also fixes the failures on Windows, which I think were caused by exactly this problem.

It may also solve the problem I was seeing with running out of file descriptors when running more tests and larger test cases.

**Review by mheinzel — APPROVED** [review]
Good to merge now.

File: src/Database/LSMTree/Internal.hs [hunk 4]
```diff
@@ -859,12 +863,13 @@ newCursor !offsetKey th = withOpenTable th $ \thEnv -> do
...
     withOpenSession cursorSession $ \_ -> do
       withTempRegistry $ \reg -> do
-        (writeBuffer, cursorRuns) <-
+        (writeBuffer, writeBufferBlobs, cursorRuns) <-
           allocTableContent reg (tableContent thEnv)
         cursorReaders <-
           allocateMaybeTemp reg
             (Readers.new hfs hbio offsetKey (Just writeBuffer) cursorRuns)
             (Readers.close hfs hbio)
+        let cursorWBB = writeBufferBlobs
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs]: Could have renamed `writeBufferBlobs` to `cursorWBB`

File: src/Database/LSMTree/Internal.hs [hunk 5]
```diff
@@ -863,13 +863,14 @@ newCursor !offsetKey th = withOpenTable th $ \thEnv -> do
...
+        (wb, wbblobs, cursorRuns) <-
           allocTableContent reg (tableContent thEnv)
         cursorReaders <-
           allocateMaybeTemp reg
-            (Readers.new hfs hbio offsetKey (Just writeBuffer) cursorRuns)
+            (Readers.new hfs hbio
+               offsetKey (Just (wb, wbblobs)) cursorRuns)
             (Readers.close hfs hbio)
-        let cursorWBB = writeBufferBlobs
+        let cursorWBB = wbblobs
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs:873]: Or here, rename `wbblobs` to `cursorWBB`

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 6]
```diff
@@ -591,8 +592,9 @@ instance Eq (Obs h a) where
       -- blob reference immediately after an update to the table, and if the SUT
       -- returns a blob, then that's okay. If both return a blob or both return
       -- an error, then those must match exactly.
-      (OEither (Right (OBlob (WrapBlob _))), OEither (Left (OId y)))
-        | Just Model.ErrBlobRefInvalidated <- cast y -> True
+      (OEither (Right (OVector vec)), OEither (Left (OId y)))
+        | Just (OBlob (WrapBlob _), _) <- V.uncons vec
+        , Just Model.ErrBlobRefInvalidated <- cast y -> True
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:597]: Why was this changed to the vector on the LHS? It means we're now only checking the first blob reference in the vector, not all of the blob references

Comment (mheinzel) [reply test/Test/Database/LSMTree/Normal/StateMachine.hs:597]: I think the situation this covers is when the model implementation of `RetrieveBlobs` returns an error, but the real successfully returns a vector. So it needs to deal with the vector.

> we're now only checking the first blob reference in the vector

We are just making sure it's a (non-empty) vector of blobs, which was probably noisy to do in a single pattern. But if the first element is a blob, I don't see how the others could be something else.

Comment (jorisdral) [reply test/Test/Database/LSMTree/Normal/StateMachine.hs:597]: You are quite right -- (my) old version was bad :P
